# eBay Kleinanzeigen

That is a car sales dataset. It contains 50,000 datapoints of used car listings. Our goal is to clean and analyze the dataset.

Who is that for? 

Let's say if an entrepreneur would want to get into the automative world and start buying/selling cars, this project will help him get familiar with a car market.

We will work with pandas, numpy, re and itertools libraries/modules on that project.

In [1]:
import pandas as pd
import numpy as np
import re
import itertools
autos = pd.read_csv('autos.csv', encoding = 'Latin-1')

In [2]:
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

You can see above all column names and first 5 rows of a dataset. 
First of all we are going to work on column names. Change all column names to lowercase and separate all words with '_' sign. So saying that in dev language, we will convert them from camelcase to snakecase.

In [3]:
copied_col = autos.columns

In [4]:
copied_col = {"dateCrawled":"date_crawled",
                 "name":"name",
                 "seller":"seller",
                 "offerType":"offer_type",   
                 "price":"price",
                 "abtest":"abtest",
                 "vehicleType":"vehicle_type",
                 "yearOfRegistration":"registration_year",   
                 "gearbox":"gearbox",
                 "powerPS":"power_ps",    
                 "model":"model",
                 "odometer":"odometer",
                 "monthOfRegistration":"registration_month",
                 "fuelType":"fuel_type",
                 "brand":"brand",
                 "notRepairedDamage":"unrepaired_damage",
                 "dateCreated":"date_created",
                 "nrOfPictures":"nr_of_pictures",
                 "postalCode":"postal_code",
                 "lastSeen":"last_seen",
                 "dateCreated":"ad_created"
             }
autos.rename(columns = copied_col, inplace=True)


In [5]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

We can notice that columns names changed successfully.

In order to begin our research, we first take a closer look at our data. We'll be using a very helpful function in pandas that will also provide us with some fundamental database statistics. This will enable us to determine whether any data cleaning or formatting is necessary.

In [6]:
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-03-08 10:40:35  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

We will investigate all common columns:

-seller 

-price 

-gearbox 

-brand 

-postal code.

To change 'price' and 'odometer' column to float type, we had to remove all others chars, as we did below. Next step is to rename the column adding km and dollar sign.


In [7]:
autos['price'] = autos['price'].str.replace("$", "").str.replace(",","").astype(float)
autos['odometer'] = autos['odometer'].str.replace("km","").str.replace(",","").astype(float)


autos.rename({"odometer":"odometer_km"}, axis = 1, inplace=True)
autos.rename({"price":"price_$"}, axis = 1, inplace = True)

In [8]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type  price_$   abtest vehicle_type  registration_year  \
0  privat    Angebot   5000.0  control          bus               2004   
1  privat    Angebot   8500.0  control    limousine               1997   
2  privat    Angebot   8990.0     test    limousine               2009   
3  privat    Angebot   4350.0  control   kleinwagen               2007   
4  privat    Angebot   1350.0     test        kombi               2003   

     gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0    manuell       158  andere     150000.0                   3       lpg   
1  automatik       286     7er     150000.0                   6    benzin   
2    manuell       102    golf      70000.0                   7    benzin   
3  automatik        71  fortwo      70000.0                   6    benzin   
4    manuell         0   focus     150000.0                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

In [9]:
autos['price_$'].value_counts()

0.0        1421
500.0       781
1500.0      734
2500.0      643
1200.0      639
           ... 
6202.0        1
18310.0       1
898.0         1
11240.0       1
789.0         1
Name: price_$, Length: 2357, dtype: int64

Looking into price columns, you can notice that 1421 prices are 0.0. We could remove the rows with 0 value, but we would be empty on half of our list. So instead it's gonna have "Unknown" value.

In [11]:
autos.loc[autos['price_$'] == 0.0, 'price_$'] = "Unknown"
autos['price_$'].value_counts()

Unknown    1421
500.0       781
1500.0      734
2500.0      643
1000.0      639
           ... 
2430.0        1
11687.0       1
38450.0       1
35222.0       1
3318.0        1
Name: price_$, Length: 2357, dtype: int64

In [12]:
autos['odometer_km'].value_counts()

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
5000.0        967
40000.0       819
30000.0       789
20000.0       784
10000.0       264
Name: odometer_km, dtype: int64

In the odometer_km column, there are 32,424 cars with 150,000 km. It looks strange, but probably it's characteristic for that specific carsale.

In [13]:
autos[['date_crawled', 'ad_created', 'last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

Project demands to show dates frequencies in percentages instead of counts, from all three columns above . We will use only year, month and a day to operate on.

In [14]:
round(autos['date_crawled'].str[:10].value_counts(normalize=True, dropna=False).sort_index(), 3)*100


2016-03-05    2.5
2016-03-06    1.4
2016-03-07    3.6
2016-03-08    3.3
2016-03-09    3.3
2016-03-10    3.2
2016-03-11    3.2
2016-03-12    3.7
2016-03-13    1.6
2016-03-14    3.7
2016-03-15    3.4
2016-03-16    3.0
2016-03-17    3.2
2016-03-18    1.3
2016-03-19    3.5
2016-03-20    3.8
2016-03-21    3.8
2016-03-22    3.3
2016-03-23    3.2
2016-03-24    2.9
2016-03-25    3.2
2016-03-26    3.2
2016-03-27    3.1
2016-03-28    3.5
2016-03-29    3.4
2016-03-30    3.4
2016-03-31    3.2
2016-04-01    3.4
2016-04-02    3.5
2016-04-03    3.9
2016-04-04    3.7
2016-04-05    1.3
2016-04-06    0.3
2016-04-07    0.1
Name: date_crawled, dtype: float64

In [15]:
autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index()*100

2015-06-11    0.002
2015-08-10    0.002
2015-09-09    0.002
2015-11-10    0.002
2015-12-05    0.002
              ...  
2016-04-03    3.892
2016-04-04    3.688
2016-04-05    1.184
2016-04-06    0.326
2016-04-07    0.128
Name: ad_created, Length: 76, dtype: float64

In [16]:
autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index()*100

2016-03-05     0.108
2016-03-06     0.442
2016-03-07     0.536
2016-03-08     0.760
2016-03-09     0.986
2016-03-10     1.076
2016-03-11     1.252
2016-03-12     2.382
2016-03-13     0.898
2016-03-14     1.280
2016-03-15     1.588
2016-03-16     1.644
2016-03-17     2.792
2016-03-18     0.742
2016-03-19     1.574
2016-03-20     2.070
2016-03-21     2.074
2016-03-22     2.158
2016-03-23     1.858
2016-03-24     1.956
2016-03-25     1.920
2016-03-26     1.696
2016-03-27     1.602
2016-03-28     2.086
2016-03-29     2.234
2016-03-30     2.484
2016-03-31     2.384
2016-04-01     2.310
2016-04-02     2.490
2016-04-03     2.536
2016-04-04     2.462
2016-04-05    12.428
2016-04-06    22.100
2016-04-07    13.092
Name: last_seen, dtype: float64

Above we introduced all 3 columns dates with percentages.

In [17]:
autos['registration_year'].describe()

count    50000.000000
mean      2005.073280
std        105.712813
min       1000.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

In registration year column, we cann see that highest year is 9999 and if there is no time traveling, it is incorrect value. Same with earliest year, no way that car could be registered at 1000 obviously. Let's stay with cars registered between 1900 and 2016.

In [18]:
autos[autos['registration_year'].between(1900, 2016)]

date_crawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offer_type price_$   abtest vehicle_type  registration_year  \
0      privat    Angebot    5000  control          bus               2004   
1      privat    Angebot    8500  control    limousine               1997   
2      privat    Angebot    8990     test    limousine               2009   
3      privat    Angebot    4350  control   kleinwagen               2007   
4      privat    Angebot    1350     test        kombi               2003   
...       ...        ...     ...      ...          ...                ...   
49995  privat    Angebot   24900  control    limousine               2011   
49996  privat    Angebot    1980  control       cabrio               1996   
49997  privat    Angebot   13200     test       cabrio               2014   
49998  privat    Angebot   22900  control        kombi               2013   
49999  privat    Angebot    1250  control    limousine               1996   

         gearbox  power_ps   model  odometer_km  registration_month fuel_type  \
0        manuell       158  andere     150000.0                   3       lpg   
1      automatik       286     7er     150000.0                   6    benzin   
2        manuell       102    golf      70000.0                   7    benzin   
3      automatik        71  fortwo      70000.0                   6    benzin   
4        manuell         0   focus     150000.0                   7    benzin   
...          ...       ...     ...          ...                 ...       ...   
49995  automatik       239      q5     100000.0                   1    diesel   
49996    manuell        75   astra     150000.0                   5    benzin   
49997  automatik        69     500       5000.0                  11    benzin   
49998    manuell       150      a3      40000.0                  11    diesel   
49999    manuell       101  vectra     150000.0                   1    benzin   

            brand unrepaired_damage           ad_created  nr_of_pictures  \
0         peugeot              nein  2016-03-26 00:00:00               0   
1             bmw              nein  2016-04-04 00:00:00               0   
2      volkswagen              nein  2016-03-26 00:00:00               0   
3           smart              nein  2016-03-12 00:00:00               0   
4            ford              nein  2016-04-01 00:00:00               0   
...           ...               ...                  ...             ...   
49995        audi              nein  2016-03-27 00:00:00               0   
49996        opel              nein  2016-03-28 00:00:00               0   
49997        fiat              nein  2016-04-02 00:00:00               0   
49998        audi              nein  2016-03-08 00:00:00               0   
49999        opel              nein  2016-03-13 00:00:00               0   

       postal_code            last_seen  
0            79588  2016-04-06 06:45:54  
1            71034  2016-04-06 14:45:08  
2            35394  2016-04-06 20:15:37  
3            33729  2016-03-15 03:16:28  
4            39218  

We choose only registered cars in years between 1900 and 2016. Only 2000 cars are with incorrect registration year so we will pop them out of our Data.

In [19]:
autos = autos[autos['registration_year'].between(1900, 2016)]

In [20]:
autos['registration_year'].describe

<bound method NDFrame.describe of 0        2004
1        1997
2        2009
3        2007
4        2003
         ... 
49995    2011
49996    1996
49997    2014
49998    2013
49999    1996
Name: registration_year, Length: 48028, dtype: int64>

We did it successfully. Let's show registration years in percentages below.

In [21]:
autos['registration_year'].value_counts(normalize=True)*100

2000    6.983426
2005    6.277588
1999    6.246356
2004    5.698759
2003    5.677938
          ...   
1939    0.002082
1927    0.002082
1929    0.002082
1948    0.002082
1952    0.002082
Name: registration_year, Length: 78, dtype: float64

#### Now we will calculate mean car price  for top 20 car brands.


In [22]:
top_cars = autos['brand'].value_counts().head(20)



In [23]:
autos['price_$'].describe()

count       48028
unique       2334
top       Unknown
freq         1335
Name: price_$, dtype: object

Because on price column we overwrote value 0 with value 'Unknown' and now we need mean of all values in that column, below there is a function which pass rows with "Unknown" value.

In [24]:
def mean_2(lista):
    avg = 0
    count = 0
    for row in lista:
        if row =='Unknown':
            pass
        else:
            avg += row
            count += 1
    final_avg = avg/count
    return final_avg


In [25]:
top_cars_brands = top_cars.index


mean_price_brand = {}

for row in top_cars_brands:
    
        avg_price = mean_2(autos.loc[autos['brand'] == row, 'price_$'])
        mean_price_brand[row] = round(avg_price)

print(mean_price_brand)

{'volkswagen': 6730, 'bmw': 8571, 'opel': 5432, 'mercedes_benz': 30829, 'audi': 9337, 'ford': 7457, 'renault': 2475, 'peugeot': 3094, 'fiat': 2814, 'seat': 4397, 'skoda': 6368, 'mazda': 4113, 'nissan': 4743, 'citroen': 45487, 'smart': 3580, 'toyota': 5167, 'sonstige_autos': 45208, 'hyundai': 5365, 'volvo': 4947, 'mini': 10613}


Above there is a list with mean price brand. The most expensive ones are Mercedes(30,829), Citroen(45487) and

In [26]:
top_6_brands_price = dict(itertools.islice(mean_price_brand.items(), 6))

In [27]:
mean_mileage_brand = {}
for row in top_cars_brands:
    
    avg_mileage = mean_2(autos.loc[autos['brand'] == row, 'odometer_km'])
    mean_mileage_brand[row] = int(avg_mileage)
print(mean_mileage_brand)

{'volkswagen': 128730, 'bmw': 132434, 'opel': 129227, 'mercedes_benz': 130860, 'audi': 129287, 'ford': 124046, 'renault': 128183, 'peugeot': 127136, 'fiat': 116553, 'seat': 121563, 'skoda': 110954, 'mazda': 124745, 'nissan': 118572, 'citroen': 119461, 'smart': 99595, 'toyota': 115709, 'sonstige_autos': 87262, 'hyundai': 106511, 'volvo': 138355, 'mini': 88602}


In [31]:
top_6_brands_mileage = dict(itertools.islice(mean_mileage_brand.items(), 6))

In [32]:
series_brands_price = pd.Series(top_6_brands_price)
print(series_brands_price)
series_brands_mileage = pd.Series(top_6_brands_mileage)
print(series_brands_mileage)

volkswagen        6730
bmw               8571
opel              5432
mercedes_benz    30829
audi              9337
ford              7457
dtype: int64
volkswagen       128730
bmw              132434
opel             129227
mercedes_benz    130860
audi             129287
ford             124046
dtype: int64


In [33]:
sb_price = pd.Series(series_brands_price)
sb_mileage = pd.Series(series_brands_mileage)
df = pd.DataFrame(sb_price, columns=)